core/server/src/main/java/alluxio/master/journal/JournalWriter.java

```diff
@@ -76,12 +97,52 @@
     mJournal = Preconditions.checkNotNull(journal);
     mJournalDirectory = mJournal.getDirectory();
     mCompletedDirectory = mJournal.getCompletedDirectory();
+    mCheckpointPath = mJournal.getCheckpointFilePath();
     mTempCheckpointPath = mJournal.getCheckpointFilePath() + ".tmp";
+    mTempBackupCheckpointPath = mJournal.getCheckpointFilePath() + ".backup.tmp";
+    mBackupCheckpointPath = mJournal.getCheckpointFilePath() + ".backup";
     mUfs = UnderFileSystem.get(mJournalDirectory);
     mMaxLogSize = Configuration.getBytes(PropertyKey.MASTER_JOURNAL_LOG_SIZE_BYTES_MAX);
   }
 
   /**
+   * Recovers the checkpoint file in case the master crashed while updating it previously.
+   *
+   * The checkpointing process goes
+   * <pre>
+   * 1. Write mTempCheckpointPath based on all completed logs
+   * 2. Rename mCheckpointPath to mTempBackupCheckpointPath
+   * 3. Rename mTempBackupCheckpointPath to mBackupCheckpointPath
+   * 4. Rename mTempCheckpointPath to mCheckpointPath
+   * 5. Delete completed logs
+   * 6. Delete mBackupCheckpointPath
+   * </pre>
+   */
+  public synchronized void recoverCheckpoint() {
+    try {
+      if (mUfs.exists(mTempBackupCheckpointPath)) {
+        // If mCheckpointPath exists, step 2 must have implemented rename as copy + delete, and
```

Comment:
The only assumption is that all three files don't exist. I've added a Precondition for this